### DISCUSSION TITLE
What is the right way to specify operations?

### DISCUSSION THREAD

**Discussion by renja-g** [api-platform/core#5420] [Q&A] [answered]
Hi, in the docs it tells me to do something like this:
```php
#[ApiResource(
    operations: [
        new Get(),
        new GetCollection()
    ]
)]
```

But when I look into the demo API it looks like this:
```php
#[GetCollection]
#[Post]
#[Get]
#[Put]
#[Patch]
#[Delete]
#[Put(
    uriTemplate: '/books/{id}/generate-cover.{_format}',
    normalizationContext: ['groups' => ['book:read', 'book:cover']],
    output: false,
    messenger: true,
)]
```

So what is the right way?
Inside the `#ApiResource` or outside?

**Comment by alanpoulain** [comment, marked as answer]
You can use both, but the core team prefers the first one (personally I prefer the second one :slightly_smiling_face:).

**Comment by renja-g** [comment]
Ahh ok, Thanks

**Comment by soyuka** [comment]
We recommend the first one, as the one with attributes under each another was introduced for backward compatibility reasons, especially that at first attributes did not support instantiating classes inside the attribute and calling `new Get` would not have been possible at that time.
